pkg/apis/core/types.go

```diff
@@ -5763,3 +5771,444 @@ type PortStatus struct {
+	// nodes by listing those where that is not the case in
+	// UnsuitableNodes.
+	//
+	// Nodes listed here will be ignored by the scheduler when selecting a
+	// node for a Pod. All other nodes are potential candidates, either
+	// because no information is available yet or because allocation might
+	// succeed.
+	//
+	// This can change, so the driver must refresh this information
+	// periodically and/or after changing resource allocation for some
+	// other ResourceClaim until a node gets selected by the scheduler.
+	UnsuitableNodes []string
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PodSchedulingList is a collection of pod scheduling objects.
+type PodSchedulingList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Items is the list of resource claims.
+	Items []PodScheduling `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ResourceClass is used by administrators to influence how resources
+// are allocated.
+type ResourceClass struct {
+	metav1.TypeMeta
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta
+
+	// DriverName determines which resource driver is to be used for
+	// allocation of a ResourceClaim that uses this class.
+	//
+	// Resource drivers have a unique name in reverse domain order
+	// (acme.example.com).
+	DriverName string
+
+	// Parameters holds arbitrary values that will be available to the
+	// driver when allocating a resource that uses this class. The driver
+	// will be able to distinguish between parameters stored here and and
+	// those stored in ResourceClaimSpec. These parameters here can only be
+	// set by cluster administrators.
+	Parameters *ResourceClassParametersReference
+
+	// Only nodes matching the selector will be considered by the scheduler
+	// when trying to find a Node that fits a Pod when that Pod uses
+	// a ResourceClaim that has not been allocated yet.
+	//
+	// Setting this field is optional. If nil, all nodes are candidates.
+	SuitableNodes *NodeSelector
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ResourceClassList is a collection of resource classes.
+type ResourceClassList struct {
+	metav1.TypeMeta
+	// Standard list metadata
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ListMeta
+
+	// Items is the list of resource classes.
+	Items []ResourceClass
+}
+
+// ResourceClassParametersReference contains enough information to let you
+// locate the parameters for a ResourceClass. The object must be
+// cluster-scoped.
```

Comment:
We postulated that for the sake of consistency with ResourceClass, which is cluster-scoped. But do we actually gain anything from imposing this restriction?

The example driver reads key/value pairs from a ConfigMap instead of definining its own CRD (simpler this way). It can currently do that for ResourceClaim parameters, but not for ResourceClass parameters, because ConfigMap isn't cluster-scoped and there's no place to put a namespace into the reference.

Real drivers might also prefer to have their parameters defined in a namespace.